cou/apps/core.py

```diff
@@ -40,3 +45,117 @@ class Octavia(OpenStackApplication):
     """
 
     wait_timeout = 30 * 60  # 30 min
+
+
+@AppFactory.register_application(["nova-compute"])
+class NovaCompute(OpenStackApplication):
+    """Nova Compute application.
+
+    Nova Compute must wait for the entire model to be idle before declaring the upgrade complete.
+    """
+
+    wait_timeout = 30 * 60  # 30 min
+    wait_for_model = True
+    force: bool = False
+
+    def upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[ApplicationUnit]]
+    ) -> list[UpgradeStep]:
+        """Upgrade steps planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate upgrade steps,
+        :type units: Optional[list[ApplicationUnit]]
+        :raises ApplicationError: If no units are passed to upgrade Nova Compute
+        :return: List of upgrade steps.
+        :rtype: list[UpgradeStep]
+        """
+        if not units:
+            raise ApplicationError("No units passed to upgrade Nova Compute.")
```

Comment:
This should not be here, it should be possible to upgrade all units, when we do not pass list of units.

Reply:
I agree that this behavior should happen for OpenStack Application, but not for NovaCompute. If no units are passed to NovaCompute is a strange situation. Maybe that means that it has no units to upgrade for a certain target.